### PR TITLE
Add all missing ES5 Array prototype functions to DefineList

### PR DESCRIPTION
**Pull request by bmomberger-bitovi**
Covers all functions explicitly listed in https://github.com/canjs/can-define/issues/81

Tests and documentation included.

### CODE REVIEW

**Comment by bmomberger-bitovi** [comment]
Serializing the list goes too deep; the callbacks to these functions should get the original items and not their serialized representations.  Even .get() with no arguments is a deep serialization, and not what we should do there. My understanding of the list.js code (before these additions) is that the comprehension functions go through `.each()` because it calls 1-argument `.get()` on each index and therefore adds observations where they may be necessary.

I think wrapper functions that call .get() and bind the callbacks properly are the way to go here.  map() and filter() could also fall under that.  Let me try a few things.

**Comment by bmomberger-bitovi** [comment]
@justinbmeyer check it out now.  Map/filter work fine with a callback wrapper in the Array functions, as do all the functions I just added.

**Comment by bmomberger-bitovi** [comment]
The linter certainly doesn't like it.  I'll slice those into arrays.

**Comment by bmomberger-bitovi** [comment]
All told, the function to proxy into Array.prototype clocks in at 30 lines, while map and filter as separate functions were 24 lines combined. I think that's acceptable.

**Comment by bmomberger-bitovi** [comment]
I started working on the ES6 ones as well ("copyWithin", "find", "findIndex", "fill", "includes", "entries", "keys") but then I realized I'd have to polyfill them, so they're on hold for now.

File: list/list.js [hunk 1]
```diff
@@ -652,6 +652,297 @@ each({
...
+	 * @param {function(item, index, list)} callback A function to call with each element of the DefineList.
+	 * The three parameters that callback gets passed are:
+	 *    - item - the element at index.
+	 *    - index - the current element of the list.
+	 *    - list - the DefineList the elements are coming from.
+	 *
+	 * The return value of `callback`, including `undefined` values are used to populated the resulting list.
+	 *
+	 * @param {Object} [thisArg] The object to use as `this` inside the callback.
+	 * @return {can-define/list/list} The list instance.
```
Comment (phillipskevin) [inline list/list.js]: Can you make it clear that these all return new lists?

File: list/list.js [hunk 2]
```diff
@@ -652,6 +652,297 @@ each({
...
+	 *   function to call with each element of the DefineList. The three parameters that callback gets passed are:
+	 *    - item - the element at index.
+	 *    - index - the current element of the list.
+	 *    - list - the DefineList the elements are coming from.
+	 *
+	 *   If `callback` returns a truthy result, `item` will be added to the result.  Otherwise, the `item` will be
+	 *   excluded.
+	 *
+	 *   @param  {Object}  thisArg  What `this` should be in the `callback`.
+	 *   @return {can-define/list/list} A `DefineList` of the same type.
```
Comment (phillipskevin) [inline list/list.js]: "new" ?

File: list/list.js [hunk 3]
```diff
@@ -652,6 +652,297 @@ each({
...
+	 * var todosAsOneObject = todos.reduce(function(todos, todo){
+	 *   todos[todo.name] = todo.complete;
+	 *   return todos;
+	 * }, {});
+	 * todosAsOneObject //-> { dishes: false, lawn: true }
+	 * ```
+	 *
+	 * @param {function(item, index, list)} callback A function to call with each element of the DefineList.
+	 * The four parameters that callback gets passed are:
+	 * 		- current - the current aggregate value of reducing over the list -- the initial value if the first iteration
```
Comment (phillipskevin) [inline list/list.js]: is this indented strangely?

File: list/list.js [hunk 4]
```diff
@@ -652,6 +652,297 @@ each({
...
+	 *    - item - the element at index.
+	 *    - index - the current element of the list.
```
Comment (phillipskevin) [inline list/list.js]: Same as above... `the index of the current element of the list`.

File: list/list.js [hunk 5]
```diff
@@ -652,6 +652,297 @@ each({
...
+	 * var todosAsOneObject = todos.reduce(function(todos, todo){
+	 *   todos[todo.name] = todo.complete;
+	 *   return todos;
+	 * }, {});
+	 * todosAsOneObject //-> { dishes: false, lawn: true }
+	 * ```
+	 *
+	 * @param {function(item, index, list)} callback A function to call with each element of the DefineList.
+	 * The four parameters that callback gets passed are:
+	 * 		- current - the current aggregate value of reducing over the list -- the initial value if the first iteration
```
Comment (phillipskevin) [inline list/list.js]: indenting again.

File: list/list.js [hunk 6]
```diff
@@ -652,6 +652,297 @@ each({
...
+	 */
+	"reduceRight": 4,
+	/**
+	 * @function can-define/list/list.prototype.every every
+	 *
+	 * Return true if every item in a list matches a predicate.
+	 *
+	 * @signature `list.every( callback [,thisArg] )`
+	 *
+	 * Filters `list` based on the return values of `callback`.  If `callback` returns truthy for every element in
```
Comment (phillipskevin) [inline list/list.js]: Maybe remove the `Filters` line... or change it to something like what MDN says:

> tests whether all elements in the array pass the test implemented by the provided function.

File: list/list.js [hunk 7]
```diff
@@ -652,6 +652,297 @@ each({
...
+	 * var aNames = names.every(function(name){
+	 *   return name[0] === "a"
+	 * });
+	 * aNames //-> false
+	 * ```
+	 *
+	 *   @param  {function(*, Number, can-define/list/list)} callback(item, index, list) A
+	 *   function to call with each element of the DefineList. The three parameters that callback gets passed are:
+	 *    - item - the element at index.
+	 *    - index - the current element of the list.
```
Comment (phillipskevin) [inline list/list.js]: Same as other `index` params.

File: list/list.js [hunk 8]
```diff
@@ -652,6 +652,297 @@ each({
...
+	 */
+	"every": 3,
+	/**
+	 * @function can-define/list/list.prototype.some some
+	 *
+	 * Return true if at least one item in a list matches a predicate.
+	 *
+	 * @signature `list.some( callback [,thisArg] )`
+	 *
+	 * Filters `list` based on the return values of `callback`.  If `callback` returns truthy for some element in
```
Comment (phillipskevin) [inline list/list.js]: Same thing... I think `Filters` is confusing.

**Pull request merged**